{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Seq\n",
    "\n",
    "Este objeto está compuesto de la secuencia misma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('CCGGGTTTA')"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq = Seq('CCGGGTTTA')\n",
    "seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ahora que esta secuencia está definida como ADN, puedes aplicar operaciones como transcribir y traducir. Los objetos seq tienen los métodos **transcribe** y **translate**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('CCGGGUUUA')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq.transcribe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('PGL')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq.translate()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Una secuencia de ARN no puede ser transcrita, solo traducida, por eso el resultado es el mismo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('CCGGUU')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_seq = Seq('CCGGUU')\n",
    "rna_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('CCGGUU')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_seq.transcribe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('PV')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_seq.translate()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Puedes también pasar de ARN a ADN usando el método **back_transcribe**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('CCGGTT')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_seq.back_transcribe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nota que la función **transcribe** reemplaza cada \"T\" con una \"U\". En biología, el proceso de transcripción parte de la cadena plantilla, haciendo un complemento reverso (TCAG → CUGA) para generar el mRNA.\n",
    "\n",
    "![seqalign](figures/transcription.png)\n",
    "\n",
    "\n",
    "En Biopython y en la bioinformática en general, se trabaja directamente con la cadena codificante porque es mucho más fácil hacer el cambio T → U. Biopython asume que tú le vas a dar esta cadena a la función. \n",
    "\n",
    "El módulo Bio.Seq también tiene funciones de **transcribe**, **back_transcribe**, **translate** y **reverse_complement** que pueden ser usados sobre los objetos Seq o cadenas:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import translate, transcribe, back_transcribe, reverse_complement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "str"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dnaseq = 'ATGGTATAA'\n",
    "type(dnaseq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MV*'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "translate(dnaseq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AUGGUAUAA'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcribe(dnaseq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AUGGUAUAA'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnaseq = transcribe(dnaseq)\n",
    "rnaseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MV*'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "translate(rnaseq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATGGTATAA'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "back_transcribe(rnaseq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Los objetos seq se comportan casi como un string, y de hecho hay algunas operaciones permitidas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('CCGGGTTAACGTA')"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq = Seq('CCGGGTTAACGTA')\n",
    "seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('CCGGG')"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCGGGTTAACGTA\n"
     ]
    }
   ],
   "source": [
    "print(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import inspect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "class Seq(object):\n",
      "    \"\"\"Read-only sequence object (essentially a string with an alphabet).\n",
      "\n",
      "    Like normal python strings, our basic sequence object is immutable.\n",
      "    This prevents you from doing my_seq[5] = \"A\" for example, but does allow\n",
      "    Seq objects to be used as dictionary keys.\n",
      "\n",
      "    The Seq object provides a number of string like methods (such as count,\n",
      "    find, split and strip), which are alphabet aware where appropriate.\n",
      "\n",
      "    In addition to the string like sequence, the Seq object has an alphabet\n",
      "    property. This is an instance of an Alphabet class from Bio.Alphabet,\n",
      "    for example generic DNA, or IUPAC DNA. This describes the type of molecule\n",
      "    (e.g. RNA, DNA, protein) and may also indicate the expected symbols\n",
      "    (letters).\n",
      "\n",
      "    The Seq object also provides some biological methods, such as complement,\n",
      "    reverse_complement, transcribe, back_transcribe and translate (which are\n",
      "    not applicable to sequences with a protein alphabet).\n",
      "    \"\"\"\n",
      "\n",
      "    def __init__(self, data, alphabet=Alphabet.generic_alphabet):\n",
      "        \"\"\"Create a Seq object.\n",
      "\n",
      "        Arguments:\n",
      "         - seq - Sequence, required (string)\n",
      "         - alphabet - Optional argument, an Alphabet object from\n",
      "           Bio.Alphabet\n",
      "\n",
      "        You will typically use Bio.SeqIO to read in sequences from files as\n",
      "        SeqRecord objects, whose sequence will be exposed as a Seq object via\n",
      "        the seq property.\n",
      "\n",
      "        However, will often want to create your own Seq objects directly:\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import IUPAC\n",
      "        >>> my_seq = Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\",\n",
      "        ...              IUPAC.protein)\n",
      "        >>> my_seq\n",
      "        Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF', IUPACProtein())\n",
      "        >>> print(my_seq)\n",
      "        MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      "        >>> my_seq.alphabet\n",
      "        IUPACProtein()\n",
      "        \"\"\"\n",
      "        # Enforce string storage\n",
      "        if not isinstance(data, basestring):\n",
      "            raise TypeError(\n",
      "                \"The sequence data given to a Seq object should \"\n",
      "                \"be a string (not another Seq object etc)\"\n",
      "            )\n",
      "        self._data = data\n",
      "        self.alphabet = alphabet  # Seq API requirement\n",
      "\n",
      "    def __repr__(self):\n",
      "        \"\"\"Return (truncated) representation of the sequence for debugging.\"\"\"\n",
      "        if self.alphabet is Alphabet.generic_alphabet:\n",
      "            # Default used, we can omit it and simplify the representation\n",
      "            a = \"\"\n",
      "        else:\n",
      "            a = \", %r\" % self.alphabet\n",
      "        if len(self) > 60:\n",
      "            # Shows the last three letters as it is often useful to see if\n",
      "            # there is a stop codon at the end of a sequence.\n",
      "            # Note total length is 54+3+3=60\n",
      "            return \"{0}('{1}...{2}'{3!s})\".format(\n",
      "                self.__class__.__name__, str(self)[:54], str(self)[-3:], a\n",
      "            )\n",
      "        else:\n",
      "            return \"{0}({1!r}{2!s})\".format(self.__class__.__name__, self._data, a)\n",
      "\n",
      "    def __str__(self):\n",
      "        \"\"\"Return the full sequence as a python string, use str(my_seq).\n",
      "\n",
      "        Note that Biopython 1.44 and earlier would give a truncated\n",
      "        version of repr(my_seq) for str(my_seq).  If you are writing code\n",
      "        which need to be backwards compatible with really old Biopython,\n",
      "        you should continue to use my_seq.tostring() as follows::\n",
      "\n",
      "            try:\n",
      "                # The old way, removed in Biopython 1.73\n",
      "                as_string = seq_obj.tostring()\n",
      "            except AttributeError:\n",
      "                # The new way, needs Biopython 1.45 or later.\n",
      "                # Don't use this on Biopython 1.44 or older as truncates\n",
      "                as_string = str(seq_obj)\n",
      "\n",
      "        \"\"\"\n",
      "        return self._data\n",
      "\n",
      "    def __hash__(self):\n",
      "        \"\"\"Hash for comparison.\n",
      "\n",
      "        See the __cmp__ documentation - this has changed from past\n",
      "        versions of Biopython!\n",
      "        \"\"\"\n",
      "        # TODO - remove this warning in a future release\n",
      "        warnings.warn(\n",
      "            \"Biopython Seq objects now use string comparison. \"\n",
      "            \"Older versions of Biopython used object comparison. \"\n",
      "            \"During this transition, please use hash(id(my_seq)) \"\n",
      "            \"or my_dict[id(my_seq)] if you want the old behaviour, \"\n",
      "            \"or use hash(str(my_seq)) or my_dict[str(my_seq)] for \"\n",
      "            \"the new string hashing behaviour.\",\n",
      "            BiopythonWarning,\n",
      "        )\n",
      "        return hash(str(self))\n",
      "\n",
      "    def __eq__(self, other):\n",
      "        \"\"\"Compare the sequence to another sequence or a string (README).\n",
      "\n",
      "        Historically comparing Seq objects has done Python object comparison.\n",
      "        After considerable discussion (keeping in mind constraints of the\n",
      "        Python language, hashes and dictionary support), Biopython now uses\n",
      "        simple string comparison (with a warning about the change).\n",
      "\n",
      "        Note that incompatible alphabets (e.g. DNA to RNA) will trigger a\n",
      "        warning.\n",
      "\n",
      "        During this transition period, please just do explicit comparisons:\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> seq1 = Seq(\"ACGT\")\n",
      "        >>> seq2 = Seq(\"ACGT\")\n",
      "        >>> id(seq1) == id(seq2)\n",
      "        False\n",
      "        >>> str(seq1) == str(seq2)\n",
      "        True\n",
      "\n",
      "        The new behaviour is to use string-like equality:\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import generic_dna\n",
      "        >>> seq1 == seq2\n",
      "        True\n",
      "        >>> seq1 == \"ACGT\"\n",
      "        True\n",
      "        >>> seq1 == Seq(\"ACGT\", generic_dna)\n",
      "        True\n",
      "\n",
      "        \"\"\"\n",
      "        if hasattr(other, \"alphabet\"):\n",
      "            # other could be a Seq or a MutableSeq\n",
      "            if not Alphabet._check_type_compatible([self.alphabet, other.alphabet]):\n",
      "                warnings.warn(\n",
      "                    \"Incompatible alphabets {0!r} and {1!r}\".format(\n",
      "                        self.alphabet, other.alphabet\n",
      "                    ),\n",
      "                    BiopythonWarning,\n",
      "                )\n",
      "        return str(self) == str(other)\n",
      "\n",
      "    def __ne__(self, other):\n",
      "        \"\"\"Implement the not-equal operand.\"\"\"\n",
      "        # Require this method for Python 2 but not needed on Python 3\n",
      "        return not self == other\n",
      "\n",
      "    def __lt__(self, other):\n",
      "        \"\"\"Implement the less-than operand.\"\"\"\n",
      "        if hasattr(other, \"alphabet\"):\n",
      "            if not Alphabet._check_type_compatible([self.alphabet, other.alphabet]):\n",
      "                warnings.warn(\n",
      "                    \"Incompatible alphabets {0!r} and {1!r}\".format(\n",
      "                        self.alphabet, other.alphabet\n",
      "                    ),\n",
      "                    BiopythonWarning,\n",
      "                )\n",
      "        if isinstance(other, (str, Seq, MutableSeq, UnknownSeq)):\n",
      "            return str(self) < str(other)\n",
      "        raise TypeError(\n",
      "            \"'<' not supported between instances of '{}' and '{}'\".format(\n",
      "                type(self).__name__, type(other).__name__\n",
      "            )\n",
      "        )\n",
      "\n",
      "    def __le__(self, other):\n",
      "        \"\"\"Implement the less-than or equal operand.\"\"\"\n",
      "        if hasattr(other, \"alphabet\"):\n",
      "            if not Alphabet._check_type_compatible([self.alphabet, other.alphabet]):\n",
      "                warnings.warn(\n",
      "                    \"Incompatible alphabets {0!r} and {1!r}\".format(\n",
      "                        self.alphabet, other.alphabet\n",
      "                    ),\n",
      "                    BiopythonWarning,\n",
      "                )\n",
      "        if isinstance(other, (str, Seq, MutableSeq, UnknownSeq)):\n",
      "            return str(self) <= str(other)\n",
      "        raise TypeError(\n",
      "            \"'<=' not supported between instances of '{}' and '{}'\".format(\n",
      "                type(self).__name__, type(other).__name__\n",
      "            )\n",
      "        )\n",
      "\n",
      "    def __gt__(self, other):\n",
      "        \"\"\"Implement the greater-than operand.\"\"\"\n",
      "        if hasattr(other, \"alphabet\"):\n",
      "            if not Alphabet._check_type_compatible([self.alphabet, other.alphabet]):\n",
      "                warnings.warn(\n",
      "                    \"Incompatible alphabets {0!r} and {1!r}\".format(\n",
      "                        self.alphabet, other.alphabet\n",
      "                    ),\n",
      "                    BiopythonWarning,\n",
      "                )\n",
      "        if isinstance(other, (str, Seq, MutableSeq, UnknownSeq)):\n",
      "            return str(self) > str(other)\n",
      "        raise TypeError(\n",
      "            \"'>' not supported between instances of '{}' and '{}'\".format(\n",
      "                type(self).__name__, type(other).__name__\n",
      "            )\n",
      "        )\n",
      "\n",
      "    def __ge__(self, other):\n",
      "        \"\"\"Implement the greater-than or equal operand.\"\"\"\n",
      "        if hasattr(other, \"alphabet\"):\n",
      "            if not Alphabet._check_type_compatible([self.alphabet, other.alphabet]):\n",
      "                warnings.warn(\n",
      "                    \"Incompatible alphabets {0!r} and {1!r}\".format(\n",
      "                        self.alphabet, other.alphabet\n",
      "                    ),\n",
      "                    BiopythonWarning,\n",
      "                )\n",
      "        if isinstance(other, (str, Seq, MutableSeq, UnknownSeq)):\n",
      "            return str(self) >= str(other)\n",
      "        raise TypeError(\n",
      "            \"'>=' not supported between instances of '{}' and '{}'\".format(\n",
      "                type(self).__name__, type(other).__name__\n",
      "            )\n",
      "        )\n",
      "\n",
      "    def __len__(self):\n",
      "        \"\"\"Return the length of the sequence, use len(my_seq).\"\"\"\n",
      "        return len(self._data)  # Seq API requirement\n",
      "\n",
      "    def __getitem__(self, index):  # Seq API requirement\n",
      "        \"\"\"Return a subsequence of single letter, use my_seq[index].\n",
      "\n",
      "        >>> my_seq = Seq('ACTCGACGTCG')\n",
      "        >>> my_seq[5]\n",
      "        'A'\n",
      "        \"\"\"\n",
      "        # Note since Python 2.0, __getslice__ is deprecated\n",
      "        # and __getitem__ is used instead.\n",
      "        # See http://docs.python.org/ref/sequence-methods.html\n",
      "        if isinstance(index, int):\n",
      "            # Return a single letter as a string\n",
      "            return self._data[index]\n",
      "        else:\n",
      "            # Return the (sub)sequence as another Seq object\n",
      "            return Seq(self._data[index], self.alphabet)\n",
      "\n",
      "    def __add__(self, other):\n",
      "        \"\"\"Add another sequence or string to this sequence.\n",
      "\n",
      "        If adding a string to a Seq, the alphabet is preserved:\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import generic_protein\n",
      "        >>> Seq(\"MELKI\", generic_protein) + \"LV\"\n",
      "        Seq('MELKILV', ProteinAlphabet())\n",
      "\n",
      "        When adding two Seq (like) objects, the alphabets are important.\n",
      "        Consider this example:\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet.IUPAC import unambiguous_dna, ambiguous_dna\n",
      "        >>> unamb_dna_seq = Seq(\"ACGT\", unambiguous_dna)\n",
      "        >>> ambig_dna_seq = Seq(\"ACRGT\", ambiguous_dna)\n",
      "        >>> unamb_dna_seq\n",
      "        Seq('ACGT', IUPACUnambiguousDNA())\n",
      "        >>> ambig_dna_seq\n",
      "        Seq('ACRGT', IUPACAmbiguousDNA())\n",
      "\n",
      "        If we add the ambiguous and unambiguous IUPAC DNA alphabets, we get\n",
      "        the more general ambiguous IUPAC DNA alphabet:\n",
      "\n",
      "        >>> unamb_dna_seq + ambig_dna_seq\n",
      "        Seq('ACGTACRGT', IUPACAmbiguousDNA())\n",
      "\n",
      "        However, if the default generic alphabet is included, the result is\n",
      "        a generic alphabet:\n",
      "\n",
      "        >>> Seq(\"\") + ambig_dna_seq\n",
      "        Seq('ACRGT')\n",
      "\n",
      "        You can't add RNA and DNA sequences:\n",
      "\n",
      "        >>> from Bio.Alphabet import generic_dna, generic_rna\n",
      "        >>> Seq(\"ACGT\", generic_dna) + Seq(\"ACGU\", generic_rna)\n",
      "        Traceback (most recent call last):\n",
      "           ...\n",
      "        TypeError: Incompatible alphabets DNAAlphabet() and RNAAlphabet()\n",
      "\n",
      "        You can't add nucleotide and protein sequences:\n",
      "\n",
      "        >>> from Bio.Alphabet import generic_dna, generic_protein\n",
      "        >>> Seq(\"ACGT\", generic_dna) + Seq(\"MELKI\", generic_protein)\n",
      "        Traceback (most recent call last):\n",
      "           ...\n",
      "        TypeError: Incompatible alphabets DNAAlphabet() and ProteinAlphabet()\n",
      "        \"\"\"\n",
      "        if hasattr(other, \"alphabet\"):\n",
      "            # other should be a Seq or a MutableSeq\n",
      "            if not Alphabet._check_type_compatible([self.alphabet, other.alphabet]):\n",
      "                raise TypeError(\n",
      "                    \"Incompatible alphabets {0!r} and {1!r}\".format(\n",
      "                        self.alphabet, other.alphabet\n",
      "                    )\n",
      "                )\n",
      "            # They should be the same sequence type (or one of them is generic)\n",
      "            a = Alphabet._consensus_alphabet([self.alphabet, other.alphabet])\n",
      "            return self.__class__(str(self) + str(other), a)\n",
      "        elif isinstance(other, basestring):\n",
      "            # other is a plain string - use the current alphabet\n",
      "            return self.__class__(str(self) + other, self.alphabet)\n",
      "        from Bio.SeqRecord import SeqRecord  # Lazy to avoid circular imports\n",
      "\n",
      "        if isinstance(other, SeqRecord):\n",
      "            # Get the SeqRecord's __radd__ to handle this\n",
      "            return NotImplemented\n",
      "        else:\n",
      "            raise TypeError\n",
      "\n",
      "    def __radd__(self, other):\n",
      "        \"\"\"Add a sequence on the left.\n",
      "\n",
      "        If adding a string to a Seq, the alphabet is preserved:\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import generic_protein\n",
      "        >>> \"LV\" + Seq(\"MELKI\", generic_protein)\n",
      "        Seq('LVMELKI', ProteinAlphabet())\n",
      "\n",
      "        Adding two Seq (like) objects is handled via the __add__ method.\n",
      "        \"\"\"\n",
      "        if hasattr(other, \"alphabet\"):\n",
      "            # other should be a Seq or a MutableSeq\n",
      "            if not Alphabet._check_type_compatible([self.alphabet, other.alphabet]):\n",
      "                raise TypeError(\n",
      "                    \"Incompatible alphabets {0!r} and {1!r}\".format(\n",
      "                        self.alphabet, other.alphabet\n",
      "                    )\n",
      "                )\n",
      "            # They should be the same sequence type (or one of them is generic)\n",
      "            a = Alphabet._consensus_alphabet([self.alphabet, other.alphabet])\n",
      "            return self.__class__(str(other) + str(self), a)\n",
      "        elif isinstance(other, basestring):\n",
      "            # other is a plain string - use the current alphabet\n",
      "            return self.__class__(other + str(self), self.alphabet)\n",
      "        else:\n",
      "            raise TypeError\n",
      "\n",
      "    def __mul__(self, other):\n",
      "        \"\"\"Multiply Seq by integer.\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import generic_dna\n",
      "        >>> Seq('ATG') * 2\n",
      "        Seq('ATGATG')\n",
      "        >>> Seq('ATG', generic_dna) * 2\n",
      "        Seq('ATGATG', DNAAlphabet())\n",
      "        \"\"\"\n",
      "        if not isinstance(other, int):\n",
      "            raise TypeError(\n",
      "                \"can't multiply {} by non-int type\".format(self.__class__.__name__)\n",
      "            )\n",
      "        return self.__class__(str(self) * other, self.alphabet)\n",
      "\n",
      "    def __rmul__(self, other):\n",
      "        \"\"\"Multiply integer by Seq.\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import generic_dna\n",
      "        >>> 2 * Seq('ATG')\n",
      "        Seq('ATGATG')\n",
      "        >>> 2 * Seq('ATG', generic_dna)\n",
      "        Seq('ATGATG', DNAAlphabet())\n",
      "        \"\"\"\n",
      "        if not isinstance(other, int):\n",
      "            raise TypeError(\n",
      "                \"can't multiply {} by non-int type\".format(self.__class__.__name__)\n",
      "            )\n",
      "        return self.__class__(str(self) * other, self.alphabet)\n",
      "\n",
      "    def __imul__(self, other):\n",
      "        \"\"\"Multiply Seq in-place.\n",
      "\n",
      "        Note although Seq is immutable, the in-place method is\n",
      "        included to match the behaviour for regular Python strings.\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import generic_dna\n",
      "        >>> seq = Seq('ATG', generic_dna)\n",
      "        >>> seq *= 2\n",
      "        >>> seq\n",
      "        Seq('ATGATG', DNAAlphabet())\n",
      "        \"\"\"\n",
      "        if not isinstance(other, int):\n",
      "            raise TypeError(\n",
      "                \"can't multiply {} by non-int type\".format(self.__class__.__name__)\n",
      "            )\n",
      "        return self.__class__(str(self) * other, self.alphabet)\n",
      "\n",
      "    def tomutable(self):  # Needed?  Or use a function?\n",
      "        \"\"\"Return the full sequence as a MutableSeq object.\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import IUPAC\n",
      "        >>> my_seq = Seq(\"MKQHKAMIVALIVICITAVVAAL\",\n",
      "        ...              IUPAC.protein)\n",
      "        >>> my_seq\n",
      "        Seq('MKQHKAMIVALIVICITAVVAAL', IUPACProtein())\n",
      "        >>> my_seq.tomutable()\n",
      "        MutableSeq('MKQHKAMIVALIVICITAVVAAL', IUPACProtein())\n",
      "\n",
      "        Note that the alphabet is preserved.\n",
      "        \"\"\"\n",
      "        return MutableSeq(str(self), self.alphabet)\n",
      "\n",
      "    def _get_seq_str_and_check_alphabet(self, other_sequence):\n",
      "        \"\"\"Convert string/Seq/MutableSeq to string, checking alphabet (PRIVATE).\n",
      "\n",
      "        For a string argument, returns the string.\n",
      "\n",
      "        For a Seq or MutableSeq, it checks the alphabet is compatible\n",
      "        (raising an exception if it isn't), and then returns a string.\n",
      "        \"\"\"\n",
      "        try:\n",
      "            other_alpha = other_sequence.alphabet\n",
      "        except AttributeError:\n",
      "            # Assume other_sequence is a string\n",
      "            return other_sequence\n",
      "\n",
      "        # Other should be a Seq or a MutableSeq\n",
      "        if not Alphabet._check_type_compatible([self.alphabet, other_alpha]):\n",
      "            raise TypeError(\n",
      "                \"Incompatible alphabets {0!r} and {1!r}\".format(\n",
      "                    self.alphabet, other_alpha\n",
      "                )\n",
      "            )\n",
      "        # Return as a string\n",
      "        return str(other_sequence)\n",
      "\n",
      "    def count(self, sub, start=0, end=sys.maxsize):\n",
      "        \"\"\"Return a non-overlapping count, like that of a python string.\n",
      "\n",
      "        This behaves like the python string method of the same name,\n",
      "        which does a non-overlapping count!\n",
      "\n",
      "        For an overlapping search use the newer count_overlap() method.\n",
      "\n",
      "        Returns an integer, the number of occurrences of substring\n",
      "        argument sub in the (sub)sequence given by [start:end].\n",
      "        Optional arguments start and end are interpreted as in slice\n",
      "        notation.\n",
      "\n",
      "        Arguments:\n",
      "         - sub - a string or another Seq object to look for\n",
      "         - start - optional integer, slice start\n",
      "         - end - optional integer, slice end\n",
      "\n",
      "        e.g.\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> my_seq = Seq(\"AAAATGA\")\n",
      "        >>> print(my_seq.count(\"A\"))\n",
      "        5\n",
      "        >>> print(my_seq.count(\"ATG\"))\n",
      "        1\n",
      "        >>> print(my_seq.count(Seq(\"AT\")))\n",
      "        1\n",
      "        >>> print(my_seq.count(\"AT\", 2, -1))\n",
      "        1\n",
      "\n",
      "        HOWEVER, please note because python strings and Seq objects (and\n",
      "        MutableSeq objects) do a non-overlapping search, this may not give\n",
      "        the answer you expect:\n",
      "\n",
      "        >>> \"AAAA\".count(\"AA\")\n",
      "        2\n",
      "        >>> print(Seq(\"AAAA\").count(\"AA\"))\n",
      "        2\n",
      "\n",
      "        An overlapping search, as implemented in .count_overlap(),\n",
      "        would give the answer as three!\n",
      "        \"\"\"\n",
      "        # If it has one, check the alphabet:\n",
      "        sub_str = self._get_seq_str_and_check_alphabet(sub)\n",
      "        return str(self).count(sub_str, start, end)\n",
      "\n",
      "    def count_overlap(self, sub, start=0, end=sys.maxsize):\n",
      "        \"\"\"Return an overlapping count.\n",
      "\n",
      "        For a non-overlapping search use the count() method.\n",
      "\n",
      "        Returns an integer, the number of occurrences of substring\n",
      "        argument sub in the (sub)sequence given by [start:end].\n",
      "        Optional arguments start and end are interpreted as in slice\n",
      "        notation.\n",
      "\n",
      "        Arguments:\n",
      "         - sub - a string or another Seq object to look for\n",
      "         - start - optional integer, slice start\n",
      "         - end - optional integer, slice end\n",
      "\n",
      "        e.g.\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> print(Seq(\"AAAA\").count_overlap(\"AA\"))\n",
      "        3\n",
      "        >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\"))\n",
      "        4\n",
      "        >>> print(Seq(\"ATATATATA\").count_overlap(\"ATA\", 3, -1))\n",
      "        1\n",
      "\n",
      "        Where substrings do not overlap, should behave the same as\n",
      "        the count() method:\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> my_seq = Seq(\"AAAATGA\")\n",
      "        >>> print(my_seq.count_overlap(\"A\"))\n",
      "        5\n",
      "        >>> my_seq.count_overlap(\"A\") == my_seq.count(\"A\")\n",
      "        True\n",
      "        >>> print(my_seq.count_overlap(\"ATG\"))\n",
      "        1\n",
      "        >>> my_seq.count_overlap(\"ATG\") == my_seq.count(\"ATG\")\n",
      "        True\n",
      "        >>> print(my_seq.count_overlap(Seq(\"AT\")))\n",
      "        1\n",
      "        >>> my_seq.count_overlap(Seq(\"AT\")) == my_seq.count(Seq(\"AT\"))\n",
      "        True\n",
      "        >>> print(my_seq.count_overlap(\"AT\", 2, -1))\n",
      "        1\n",
      "        >>> my_seq.count_overlap(\"AT\", 2, -1) == my_seq.count(\"AT\", 2, -1)\n",
      "        True\n",
      "\n",
      "        HOWEVER, do not use this method for such cases because the\n",
      "        count() method is much for efficient.\n",
      "        \"\"\"\n",
      "        sub_str = self._get_seq_str_and_check_alphabet(sub)\n",
      "        self_str = str(self)\n",
      "        overlap_count = 0\n",
      "        while True:\n",
      "            start = self_str.find(sub_str, start, end) + 1\n",
      "            if start != 0:\n",
      "                overlap_count += 1\n",
      "            else:\n",
      "                return overlap_count\n",
      "\n",
      "    def __contains__(self, char):\n",
      "        \"\"\"Implement the 'in' keyword, like a python string.\n",
      "\n",
      "        e.g.\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import generic_dna, generic_rna, generic_protein\n",
      "        >>> my_dna = Seq(\"ATATGAAATTTGAAAA\", generic_dna)\n",
      "        >>> \"AAA\" in my_dna\n",
      "        True\n",
      "        >>> Seq(\"AAA\") in my_dna\n",
      "        True\n",
      "        >>> Seq(\"AAA\", generic_dna) in my_dna\n",
      "        True\n",
      "\n",
      "        Like other Seq methods, this will raise a type error if another Seq\n",
      "        (or Seq like) object with an incompatible alphabet is used:\n",
      "\n",
      "        >>> Seq(\"AAA\", generic_rna) in my_dna\n",
      "        Traceback (most recent call last):\n",
      "           ...\n",
      "        TypeError: Incompatible alphabets DNAAlphabet() and RNAAlphabet()\n",
      "        >>> Seq(\"AAA\", generic_protein) in my_dna\n",
      "        Traceback (most recent call last):\n",
      "           ...\n",
      "        TypeError: Incompatible alphabets DNAAlphabet() and ProteinAlphabet()\n",
      "        \"\"\"\n",
      "        # If it has one, check the alphabet:\n",
      "        sub_str = self._get_seq_str_and_check_alphabet(char)\n",
      "        return sub_str in str(self)\n",
      "\n",
      "    def find(self, sub, start=0, end=sys.maxsize):\n",
      "        \"\"\"Find method, like that of a python string.\n",
      "\n",
      "        This behaves like the python string method of the same name.\n",
      "\n",
      "        Returns an integer, the index of the first occurrence of substring\n",
      "        argument sub in the (sub)sequence given by [start:end].\n",
      "\n",
      "        Arguments:\n",
      "         - sub - a string or another Seq object to look for\n",
      "         - start - optional integer, slice start\n",
      "         - end - optional integer, slice end\n",
      "\n",
      "        Returns -1 if the subsequence is NOT found.\n",
      "\n",
      "        e.g. Locating the first typical start codon, AUG, in an RNA sequence:\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      "        >>> my_rna.find(\"AUG\")\n",
      "        3\n",
      "        \"\"\"\n",
      "        # If it has one, check the alphabet:\n",
      "        sub_str = self._get_seq_str_and_check_alphabet(sub)\n",
      "        return str(self).find(sub_str, start, end)\n",
      "\n",
      "    def rfind(self, sub, start=0, end=sys.maxsize):\n",
      "        \"\"\"Find from right method, like that of a python string.\n",
      "\n",
      "        This behaves like the python string method of the same name.\n",
      "\n",
      "        Returns an integer, the index of the last (right most) occurrence of\n",
      "        substring argument sub in the (sub)sequence given by [start:end].\n",
      "\n",
      "        Arguments:\n",
      "         - sub - a string or another Seq object to look for\n",
      "         - start - optional integer, slice start\n",
      "         - end - optional integer, slice end\n",
      "\n",
      "        Returns -1 if the subsequence is NOT found.\n",
      "\n",
      "        e.g. Locating the last typical start codon, AUG, in an RNA sequence:\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      "        >>> my_rna.rfind(\"AUG\")\n",
      "        15\n",
      "        \"\"\"\n",
      "        # If it has one, check the alphabet:\n",
      "        sub_str = self._get_seq_str_and_check_alphabet(sub)\n",
      "        return str(self).rfind(sub_str, start, end)\n",
      "\n",
      "    def index(self, sub, start=0, end=sys.maxsize):\n",
      "        \"\"\"Like find() but raise ValueError when the substring is not found.\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      "        >>> my_rna.find(\"T\")\n",
      "        -1\n",
      "        >>> my_rna.index(\"T\")\n",
      "        Traceback (most recent call last):\n",
      "                   ...\n",
      "        ValueError: substring not found...\n",
      "        \"\"\"\n",
      "        # If it has one, check the alphabet:\n",
      "        sub_str = self._get_seq_str_and_check_alphabet(sub)\n",
      "        return str(self).index(sub_str, start, end)\n",
      "\n",
      "    def rindex(self, sub, start=0, end=sys.maxsize):\n",
      "        \"\"\"Like rfind() but raise ValueError when the substring is not found.\"\"\"\n",
      "        # If it has one, check the alphabet:\n",
      "        sub_str = self._get_seq_str_and_check_alphabet(sub)\n",
      "        return str(self).rindex(sub_str, start, end)\n",
      "\n",
      "    def startswith(self, prefix, start=0, end=sys.maxsize):\n",
      "        \"\"\"Return True if the Seq starts with the given prefix, False otherwise.\n",
      "\n",
      "        This behaves like the python string method of the same name.\n",
      "\n",
      "        Return True if the sequence starts with the specified prefix\n",
      "        (a string or another Seq object), False otherwise.\n",
      "        With optional start, test sequence beginning at that position.\n",
      "        With optional end, stop comparing sequence at that position.\n",
      "        prefix can also be a tuple of strings to try.  e.g.\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      "        >>> my_rna.startswith(\"GUC\")\n",
      "        True\n",
      "        >>> my_rna.startswith(\"AUG\")\n",
      "        False\n",
      "        >>> my_rna.startswith(\"AUG\", 3)\n",
      "        True\n",
      "        >>> my_rna.startswith((\"UCC\", \"UCA\", \"UCG\"), 1)\n",
      "        True\n",
      "        \"\"\"\n",
      "        # If it has one, check the alphabet:\n",
      "        if isinstance(prefix, tuple):\n",
      "            prefix_strs = tuple(self._get_seq_str_and_check_alphabet(p) for p in prefix)\n",
      "            return str(self).startswith(prefix_strs, start, end)\n",
      "        else:\n",
      "            prefix_str = self._get_seq_str_and_check_alphabet(prefix)\n",
      "            return str(self).startswith(prefix_str, start, end)\n",
      "\n",
      "    def endswith(self, suffix, start=0, end=sys.maxsize):\n",
      "        \"\"\"Return True if the Seq ends with the given suffix, False otherwise.\n",
      "\n",
      "        This behaves like the python string method of the same name.\n",
      "\n",
      "        Return True if the sequence ends with the specified suffix\n",
      "        (a string or another Seq object), False otherwise.\n",
      "        With optional start, test sequence beginning at that position.\n",
      "        With optional end, stop comparing sequence at that position.\n",
      "        suffix can also be a tuple of strings to try.  e.g.\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      "        >>> my_rna.endswith(\"UUG\")\n",
      "        True\n",
      "        >>> my_rna.endswith(\"AUG\")\n",
      "        False\n",
      "        >>> my_rna.endswith(\"AUG\", 0, 18)\n",
      "        True\n",
      "        >>> my_rna.endswith((\"UCC\", \"UCA\", \"UUG\"))\n",
      "        True\n",
      "        \"\"\"\n",
      "        # If it has one, check the alphabet:\n",
      "        if isinstance(suffix, tuple):\n",
      "            suffix_strs = tuple(self._get_seq_str_and_check_alphabet(p) for p in suffix)\n",
      "            return str(self).endswith(suffix_strs, start, end)\n",
      "        else:\n",
      "            suffix_str = self._get_seq_str_and_check_alphabet(suffix)\n",
      "            return str(self).endswith(suffix_str, start, end)\n",
      "\n",
      "    def split(self, sep=None, maxsplit=-1):\n",
      "        \"\"\"Split method, like that of a python string.\n",
      "\n",
      "        This behaves like the python string method of the same name.\n",
      "\n",
      "        Return a list of the 'words' in the string (as Seq objects),\n",
      "        using sep as the delimiter string.  If maxsplit is given, at\n",
      "        most maxsplit splits are done.  If maxsplit is omitted, all\n",
      "        splits are made.\n",
      "\n",
      "        Following the python string method, sep will by default be any\n",
      "        white space (tabs, spaces, newlines) but this is unlikely to\n",
      "        apply to biological sequences.\n",
      "\n",
      "        e.g.\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> my_rna = Seq(\"GUCAUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAGUUG\")\n",
      "        >>> my_aa = my_rna.translate()\n",
      "        >>> my_aa\n",
      "        Seq('VMAIVMGR*KGAR*L', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      "        >>> for pep in my_aa.split(\"*\"):\n",
      "        ...     pep\n",
      "        Seq('VMAIVMGR', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      "        Seq('KGAR', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      "        Seq('L', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      "        >>> for pep in my_aa.split(\"*\", 1):\n",
      "        ...     pep\n",
      "        Seq('VMAIVMGR', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      "        Seq('KGAR*L', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      "\n",
      "        See also the rsplit method:\n",
      "\n",
      "        >>> for pep in my_aa.rsplit(\"*\", 1):\n",
      "        ...     pep\n",
      "        Seq('VMAIVMGR*KGAR', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      "        Seq('L', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      "        \"\"\"\n",
      "        # If it has one, check the alphabet:\n",
      "        sep_str = self._get_seq_str_and_check_alphabet(sep)\n",
      "        # TODO - If the sep is the defined stop symbol, or gap char,\n",
      "        # should we adjust the alphabet?\n",
      "        return [Seq(part, self.alphabet) for part in str(self).split(sep_str, maxsplit)]\n",
      "\n",
      "    def rsplit(self, sep=None, maxsplit=-1):\n",
      "        \"\"\"Do a right split method, like that of a python string.\n",
      "\n",
      "        This behaves like the python string method of the same name.\n",
      "\n",
      "        Return a list of the 'words' in the string (as Seq objects),\n",
      "        using sep as the delimiter string.  If maxsplit is given, at\n",
      "        most maxsplit splits are done COUNTING FROM THE RIGHT.\n",
      "        If maxsplit is omitted, all splits are made.\n",
      "\n",
      "        Following the python string method, sep will by default be any\n",
      "        white space (tabs, spaces, newlines) but this is unlikely to\n",
      "        apply to biological sequences.\n",
      "\n",
      "        e.g. print(my_seq.rsplit(\"*\",1))\n",
      "\n",
      "        See also the split method.\n",
      "        \"\"\"\n",
      "        # If it has one, check the alphabet:\n",
      "        sep_str = self._get_seq_str_and_check_alphabet(sep)\n",
      "        return [\n",
      "            Seq(part, self.alphabet) for part in str(self).rsplit(sep_str, maxsplit)\n",
      "        ]\n",
      "\n",
      "    def strip(self, chars=None):\n",
      "        \"\"\"Return a new Seq object with leading and trailing ends stripped.\n",
      "\n",
      "        This behaves like the python string method of the same name.\n",
      "\n",
      "        Optional argument chars defines which characters to remove.  If\n",
      "        omitted or None (default) then as for the python string method,\n",
      "        this defaults to removing any white space.\n",
      "\n",
      "        e.g. print(my_seq.strip(\"-\"))\n",
      "\n",
      "        See also the lstrip and rstrip methods.\n",
      "        \"\"\"\n",
      "        # If it has one, check the alphabet:\n",
      "        strip_str = self._get_seq_str_and_check_alphabet(chars)\n",
      "        return Seq(str(self).strip(strip_str), self.alphabet)\n",
      "\n",
      "    def lstrip(self, chars=None):\n",
      "        \"\"\"Return a new Seq object with leading (left) end stripped.\n",
      "\n",
      "        This behaves like the python string method of the same name.\n",
      "\n",
      "        Optional argument chars defines which characters to remove.  If\n",
      "        omitted or None (default) then as for the python string method,\n",
      "        this defaults to removing any white space.\n",
      "\n",
      "        e.g. print(my_seq.lstrip(\"-\"))\n",
      "\n",
      "        See also the strip and rstrip methods.\n",
      "        \"\"\"\n",
      "        # If it has one, check the alphabet:\n",
      "        strip_str = self._get_seq_str_and_check_alphabet(chars)\n",
      "        return Seq(str(self).lstrip(strip_str), self.alphabet)\n",
      "\n",
      "    def rstrip(self, chars=None):\n",
      "        \"\"\"Return a new Seq object with trailing (right) end stripped.\n",
      "\n",
      "        This behaves like the python string method of the same name.\n",
      "\n",
      "        Optional argument chars defines which characters to remove.  If\n",
      "        omitted or None (default) then as for the python string method,\n",
      "        this defaults to removing any white space.\n",
      "\n",
      "        e.g. Removing a nucleotide sequence's polyadenylation (poly-A tail):\n",
      "\n",
      "        >>> from Bio.Alphabet import IUPAC\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> my_seq = Seq(\"CGGTACGCTTATGTCACGTAGAAAAAA\", IUPAC.unambiguous_dna)\n",
      "        >>> my_seq\n",
      "        Seq('CGGTACGCTTATGTCACGTAGAAAAAA', IUPACUnambiguousDNA())\n",
      "        >>> my_seq.rstrip(\"A\")\n",
      "        Seq('CGGTACGCTTATGTCACGTAG', IUPACUnambiguousDNA())\n",
      "\n",
      "        See also the strip and lstrip methods.\n",
      "        \"\"\"\n",
      "        # If it has one, check the alphabet:\n",
      "        strip_str = self._get_seq_str_and_check_alphabet(chars)\n",
      "        return Seq(str(self).rstrip(strip_str), self.alphabet)\n",
      "\n",
      "    def upper(self):\n",
      "        \"\"\"Return an upper case copy of the sequence.\n",
      "\n",
      "        >>> from Bio.Alphabet import HasStopCodon, generic_protein\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> my_seq = Seq(\"VHLTPeeK*\", HasStopCodon(generic_protein))\n",
      "        >>> my_seq\n",
      "        Seq('VHLTPeeK*', HasStopCodon(ProteinAlphabet(), '*'))\n",
      "        >>> my_seq.lower()\n",
      "        Seq('vhltpeek*', HasStopCodon(ProteinAlphabet(), '*'))\n",
      "        >>> my_seq.upper()\n",
      "        Seq('VHLTPEEK*', HasStopCodon(ProteinAlphabet(), '*'))\n",
      "\n",
      "        This will adjust the alphabet if required. See also the lower method.\n",
      "        \"\"\"\n",
      "        return Seq(str(self).upper(), self.alphabet._upper())\n",
      "\n",
      "    def lower(self):\n",
      "        \"\"\"Return a lower case copy of the sequence.\n",
      "\n",
      "        This will adjust the alphabet if required. Note that the IUPAC\n",
      "        alphabets are upper case only, and thus a generic alphabet must be\n",
      "        substituted.\n",
      "\n",
      "        >>> from Bio.Alphabet import Gapped, generic_dna\n",
      "        >>> from Bio.Alphabet import IUPAC\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> my_seq = Seq(\"CGGTACGCTTATGTCACGTAG*AAAAAA\",\n",
      "        ...          Gapped(IUPAC.unambiguous_dna, \"*\"))\n",
      "        >>> my_seq\n",
      "        Seq('CGGTACGCTTATGTCACGTAG*AAAAAA', Gapped(IUPACUnambiguousDNA(), '*'))\n",
      "        >>> my_seq.lower()\n",
      "        Seq('cggtacgcttatgtcacgtag*aaaaaa', Gapped(DNAAlphabet(), '*'))\n",
      "\n",
      "        See also the upper method.\n",
      "        \"\"\"\n",
      "        return Seq(str(self).lower(), self.alphabet._lower())\n",
      "\n",
      "    def encode(self, encoding=\"utf-8\", errors=\"strict\"):\n",
      "        \"\"\"Return an encoded version of the sequence as a bytes object.\n",
      "\n",
      "        The Seq object aims to match the interfact of a Python string.\n",
      "        (This means on Python 3 it acts like a unicode string.)\n",
      "\n",
      "        This is essentially to save you doing str(my_seq).encode() when\n",
      "        you need a bytes string, for example for computing a hash:\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> Seq(\"ACGT\").encode(\"ascii\")\n",
      "        b'ACGT'\n",
      "        \"\"\"\n",
      "        return str(self).encode(encoding, errors)\n",
      "\n",
      "    def complement(self):\n",
      "        \"\"\"Return the complement sequence by creating a new Seq object.\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import IUPAC\n",
      "        >>> my_dna = Seq(\"CCCCCGATAG\", IUPAC.unambiguous_dna)\n",
      "        >>> my_dna\n",
      "        Seq('CCCCCGATAG', IUPACUnambiguousDNA())\n",
      "        >>> my_dna.complement()\n",
      "        Seq('GGGGGCTATC', IUPACUnambiguousDNA())\n",
      "\n",
      "        You can of course used mixed case sequences,\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import generic_dna\n",
      "        >>> my_dna = Seq(\"CCCCCgatA-GD\", generic_dna)\n",
      "        >>> my_dna\n",
      "        Seq('CCCCCgatA-GD', DNAAlphabet())\n",
      "        >>> my_dna.complement()\n",
      "        Seq('GGGGGctaT-CH', DNAAlphabet())\n",
      "\n",
      "        Note in the above example, ambiguous character D denotes\n",
      "        G, A or T so its complement is H (for C, T or A).\n",
      "\n",
      "        Trying to complement a protein sequence raises an exception.\n",
      "\n",
      "        >>> my_protein = Seq(\"MAIVMGR\", IUPAC.protein)\n",
      "        >>> my_protein.complement()\n",
      "        Traceback (most recent call last):\n",
      "           ...\n",
      "        ValueError: Proteins do not have complements!\n",
      "        \"\"\"\n",
      "        base = Alphabet._get_base_alphabet(self.alphabet)\n",
      "        if isinstance(base, Alphabet.ProteinAlphabet):\n",
      "            raise ValueError(\"Proteins do not have complements!\")\n",
      "        if isinstance(base, Alphabet.DNAAlphabet):\n",
      "            ttable = _dna_complement_table\n",
      "        elif isinstance(base, Alphabet.RNAAlphabet):\n",
      "            ttable = _rna_complement_table\n",
      "        elif (\"U\" in self._data or \"u\" in self._data) and (\n",
      "            \"T\" in self._data or \"t\" in self._data\n",
      "        ):\n",
      "            # TODO - Handle this cleanly?\n",
      "            raise ValueError(\"Mixed RNA/DNA found\")\n",
      "        elif \"U\" in self._data or \"u\" in self._data:\n",
      "            ttable = _rna_complement_table\n",
      "        else:\n",
      "            ttable = _dna_complement_table\n",
      "        # Much faster on really long sequences than the previous loop based\n",
      "        # one. Thanks to Michael Palmer, University of Waterloo.\n",
      "        return Seq(str(self).translate(ttable), self.alphabet)\n",
      "\n",
      "    def reverse_complement(self):\n",
      "        \"\"\"Return the reverse complement sequence by creating a new Seq object.\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import IUPAC\n",
      "        >>> my_dna = Seq(\"CCCCCGATAGNR\", IUPAC.ambiguous_dna)\n",
      "        >>> my_dna\n",
      "        Seq('CCCCCGATAGNR', IUPACAmbiguousDNA())\n",
      "        >>> my_dna.reverse_complement()\n",
      "        Seq('YNCTATCGGGGG', IUPACAmbiguousDNA())\n",
      "\n",
      "        Note in the above example, since R = G or A, its complement\n",
      "        is Y (which denotes C or T).\n",
      "\n",
      "        You can of course used mixed case sequences,\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import generic_dna\n",
      "        >>> my_dna = Seq(\"CCCCCgatA-G\", generic_dna)\n",
      "        >>> my_dna\n",
      "        Seq('CCCCCgatA-G', DNAAlphabet())\n",
      "        >>> my_dna.reverse_complement()\n",
      "        Seq('C-TatcGGGGG', DNAAlphabet())\n",
      "\n",
      "        Trying to complement a protein sequence raises an exception:\n",
      "\n",
      "        >>> my_protein = Seq(\"MAIVMGR\", IUPAC.protein)\n",
      "        >>> my_protein.reverse_complement()\n",
      "        Traceback (most recent call last):\n",
      "           ...\n",
      "        ValueError: Proteins do not have complements!\n",
      "        \"\"\"\n",
      "        # Use -1 stride/step to reverse the complement\n",
      "        return self.complement()[::-1]\n",
      "\n",
      "    def transcribe(self):\n",
      "        \"\"\"Return the RNA sequence from a DNA sequence by creating a new Seq object.\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import IUPAC\n",
      "        >>> coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\",\n",
      "        ...                  IUPAC.unambiguous_dna)\n",
      "        >>> coding_dna\n",
      "        Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG', IUPACUnambiguousDNA())\n",
      "        >>> coding_dna.transcribe()\n",
      "        Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG', IUPACUnambiguousRNA())\n",
      "\n",
      "        Trying to transcribe a protein or RNA sequence raises an exception:\n",
      "\n",
      "        >>> my_protein = Seq(\"MAIVMGR\", IUPAC.protein)\n",
      "        >>> my_protein.transcribe()\n",
      "        Traceback (most recent call last):\n",
      "           ...\n",
      "        ValueError: Proteins cannot be transcribed!\n",
      "        \"\"\"\n",
      "        base = Alphabet._get_base_alphabet(self.alphabet)\n",
      "        if isinstance(base, Alphabet.ProteinAlphabet):\n",
      "            raise ValueError(\"Proteins cannot be transcribed!\")\n",
      "        if isinstance(base, Alphabet.RNAAlphabet):\n",
      "            raise ValueError(\"RNA cannot be transcribed!\")\n",
      "\n",
      "        if self.alphabet == IUPAC.unambiguous_dna:\n",
      "            alphabet = IUPAC.unambiguous_rna\n",
      "        elif self.alphabet == IUPAC.ambiguous_dna:\n",
      "            alphabet = IUPAC.ambiguous_rna\n",
      "        else:\n",
      "            alphabet = Alphabet.generic_rna\n",
      "        return Seq(str(self).replace(\"T\", \"U\").replace(\"t\", \"u\"), alphabet)\n",
      "\n",
      "    def back_transcribe(self):\n",
      "        \"\"\"Return the DNA sequence from an RNA sequence by creating a new Seq object.\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import IUPAC\n",
      "        >>> messenger_rna = Seq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\",\n",
      "        ...                     IUPAC.unambiguous_rna)\n",
      "        >>> messenger_rna\n",
      "        Seq('AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG', IUPACUnambiguousRNA())\n",
      "        >>> messenger_rna.back_transcribe()\n",
      "        Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG', IUPACUnambiguousDNA())\n",
      "\n",
      "        Trying to back-transcribe a protein or DNA sequence raises an\n",
      "        exception:\n",
      "\n",
      "        >>> my_protein = Seq(\"MAIVMGR\", IUPAC.protein)\n",
      "        >>> my_protein.back_transcribe()\n",
      "        Traceback (most recent call last):\n",
      "           ...\n",
      "        ValueError: Proteins cannot be back transcribed!\n",
      "        \"\"\"\n",
      "        base = Alphabet._get_base_alphabet(self.alphabet)\n",
      "        if isinstance(base, Alphabet.ProteinAlphabet):\n",
      "            raise ValueError(\"Proteins cannot be back transcribed!\")\n",
      "        if isinstance(base, Alphabet.DNAAlphabet):\n",
      "            raise ValueError(\"DNA cannot be back transcribed!\")\n",
      "\n",
      "        if self.alphabet == IUPAC.unambiguous_rna:\n",
      "            alphabet = IUPAC.unambiguous_dna\n",
      "        elif self.alphabet == IUPAC.ambiguous_rna:\n",
      "            alphabet = IUPAC.ambiguous_dna\n",
      "        else:\n",
      "            alphabet = Alphabet.generic_dna\n",
      "        return Seq(str(self).replace(\"U\", \"T\").replace(\"u\", \"t\"), alphabet)\n",
      "\n",
      "    def translate(\n",
      "        self, table=\"Standard\", stop_symbol=\"*\", to_stop=False, cds=False, gap=None\n",
      "    ):\n",
      "        \"\"\"Turn a nucleotide sequence into a protein sequence by creating a new Seq object.\n",
      "\n",
      "        This method will translate DNA or RNA sequences, and those with a\n",
      "        nucleotide or generic alphabet.  Trying to translate a protein\n",
      "        sequence raises an exception.\n",
      "\n",
      "        Arguments:\n",
      "         - table - Which codon table to use?  This can be either a name\n",
      "           (string), an NCBI identifier (integer), or a CodonTable\n",
      "           object (useful for non-standard genetic codes).  This\n",
      "           defaults to the \"Standard\" table.\n",
      "         - stop_symbol - Single character string, what to use for\n",
      "           terminators.  This defaults to the asterisk, \"*\".\n",
      "         - to_stop - Boolean, defaults to False meaning do a full\n",
      "           translation continuing on past any stop codons (translated as the\n",
      "           specified stop_symbol).  If True, translation is terminated at\n",
      "           the first in frame stop codon (and the stop_symbol is not\n",
      "           appended to the returned protein sequence).\n",
      "         - cds - Boolean, indicates this is a complete CDS.  If True,\n",
      "           this checks the sequence starts with a valid alternative start\n",
      "           codon (which will be translated as methionine, M), that the\n",
      "           sequence length is a multiple of three, and that there is a\n",
      "           single in frame stop codon at the end (this will be excluded\n",
      "           from the protein sequence, regardless of the to_stop option).\n",
      "           If these tests fail, an exception is raised.\n",
      "         - gap - Single character string to denote symbol used for gaps.\n",
      "           It will try to guess the gap character from the alphabet.\n",
      "\n",
      "        e.g. Using the standard table:\n",
      "\n",
      "        >>> coding_dna = Seq(\"GTGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
      "        >>> coding_dna.translate()\n",
      "        Seq('VAIVMGR*KGAR*', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      "        >>> coding_dna.translate(stop_symbol=\"@\")\n",
      "        Seq('VAIVMGR@KGAR@', HasStopCodon(ExtendedIUPACProtein(), '@'))\n",
      "        >>> coding_dna.translate(to_stop=True)\n",
      "        Seq('VAIVMGR', ExtendedIUPACProtein())\n",
      "\n",
      "        Now using NCBI table 2, where TGA is not a stop codon:\n",
      "\n",
      "        >>> coding_dna.translate(table=2)\n",
      "        Seq('VAIVMGRWKGAR*', HasStopCodon(ExtendedIUPACProtein(), '*'))\n",
      "        >>> coding_dna.translate(table=2, to_stop=True)\n",
      "        Seq('VAIVMGRWKGAR', ExtendedIUPACProtein())\n",
      "\n",
      "        In fact, GTG is an alternative start codon under NCBI table 2, meaning\n",
      "        this sequence could be a complete CDS:\n",
      "\n",
      "        >>> coding_dna.translate(table=2, cds=True)\n",
      "        Seq('MAIVMGRWKGAR', ExtendedIUPACProtein())\n",
      "\n",
      "        It isn't a valid CDS under NCBI table 1, due to both the start codon\n",
      "        and also the in frame stop codons:\n",
      "\n",
      "        >>> coding_dna.translate(table=1, cds=True)\n",
      "        Traceback (most recent call last):\n",
      "            ...\n",
      "        Bio.Data.CodonTable.TranslationError: First codon 'GTG' is not a start codon\n",
      "\n",
      "        If the sequence has no in-frame stop codon, then the to_stop argument\n",
      "        has no effect:\n",
      "\n",
      "        >>> coding_dna2 = Seq(\"TTGGCCATTGTAATGGGCCGC\")\n",
      "        >>> coding_dna2.translate()\n",
      "        Seq('LAIVMGR', ExtendedIUPACProtein())\n",
      "        >>> coding_dna2.translate(to_stop=True)\n",
      "        Seq('LAIVMGR', ExtendedIUPACProtein())\n",
      "\n",
      "        When translating gapped sequences, the gap character is inferred from\n",
      "        the alphabet:\n",
      "\n",
      "        >>> from Bio.Alphabet import Gapped\n",
      "        >>> coding_dna3 = Seq(\"GTG---GCCATT\", Gapped(IUPAC.unambiguous_dna))\n",
      "        >>> coding_dna3.translate()\n",
      "        Seq('V-AI', Gapped(ExtendedIUPACProtein(), '-'))\n",
      "\n",
      "        It is possible to pass the gap character when the alphabet is missing:\n",
      "\n",
      "        >>> coding_dna4 = Seq(\"GTG---GCCATT\")\n",
      "        >>> coding_dna4.translate(gap='-')\n",
      "        Seq('V-AI', Gapped(ExtendedIUPACProtein(), '-'))\n",
      "\n",
      "        NOTE - Ambiguous codons like \"TAN\" or \"NNN\" could be an amino acid\n",
      "        or a stop codon.  These are translated as \"X\".  Any invalid codon\n",
      "        (e.g. \"TA?\" or \"T-A\") will throw a TranslationError.\n",
      "\n",
      "        NOTE - This does NOT behave like the python string's translate\n",
      "        method.  For that use str(my_seq).translate(...) instead.\n",
      "        \"\"\"\n",
      "        if isinstance(table, str) and len(table) == 256:\n",
      "            raise ValueError(\n",
      "                \"The Seq object translate method DOES NOT take \"\n",
      "                \"a 256 character string mapping table like \"\n",
      "                \"the python string object's translate method. \"\n",
      "                \"Use str(my_seq).translate(...) instead.\"\n",
      "            )\n",
      "        if isinstance(\n",
      "            Alphabet._get_base_alphabet(self.alphabet), Alphabet.ProteinAlphabet\n",
      "        ):\n",
      "            raise ValueError(\"Proteins cannot be translated!\")\n",
      "        try:\n",
      "            table_id = int(table)\n",
      "        except ValueError:\n",
      "            # Assume its a table name\n",
      "            if self.alphabet == IUPAC.unambiguous_dna:\n",
      "                # Will use standard IUPAC protein alphabet, no need for X\n",
      "                codon_table = CodonTable.unambiguous_dna_by_name[table]\n",
      "            elif self.alphabet == IUPAC.unambiguous_rna:\n",
      "                # Will use standard IUPAC protein alphabet, no need for X\n",
      "                codon_table = CodonTable.unambiguous_rna_by_name[table]\n",
      "            else:\n",
      "                # This will use the extended IUPAC protein alphabet with X etc.\n",
      "                # The same table can be used for RNA or DNA (we use this for\n",
      "                # translating strings).\n",
      "                codon_table = CodonTable.ambiguous_generic_by_name[table]\n",
      "        except (AttributeError, TypeError):\n",
      "            # Assume its a CodonTable object\n",
      "            if isinstance(table, CodonTable.CodonTable):\n",
      "                codon_table = table\n",
      "            else:\n",
      "                raise ValueError(\"Bad table argument\")\n",
      "        else:\n",
      "            # Assume its a table ID\n",
      "            if self.alphabet == IUPAC.unambiguous_dna:\n",
      "                # Will use standard IUPAC protein alphabet, no need for X\n",
      "                codon_table = CodonTable.unambiguous_dna_by_id[table_id]\n",
      "            elif self.alphabet == IUPAC.unambiguous_rna:\n",
      "                # Will use standard IUPAC protein alphabet, no need for X\n",
      "                codon_table = CodonTable.unambiguous_rna_by_id[table_id]\n",
      "            else:\n",
      "                # This will use the extended IUPAC protein alphabet with X etc.\n",
      "                # The same table can be used for RNA or DNA (we use this for\n",
      "                # translating strings).\n",
      "                codon_table = CodonTable.ambiguous_generic_by_id[table_id]\n",
      "\n",
      "        # Deal with gaps for translation\n",
      "        if hasattr(self.alphabet, \"gap_char\"):\n",
      "            if not gap:\n",
      "                gap = self.alphabet.gap_char\n",
      "            elif gap != self.alphabet.gap_char:\n",
      "                raise ValueError(\n",
      "                    \"Gap {0!r} does not match {1!r} from alphabet\".format(\n",
      "                        gap, self.alphabet.gap_char\n",
      "                    )\n",
      "                )\n",
      "\n",
      "        protein = _translate_str(\n",
      "            str(self), codon_table, stop_symbol, to_stop, cds, gap=gap\n",
      "        )\n",
      "\n",
      "        if gap and gap in protein:\n",
      "            alphabet = Alphabet.Gapped(codon_table.protein_alphabet, gap)\n",
      "        else:\n",
      "            alphabet = codon_table.protein_alphabet\n",
      "\n",
      "        if stop_symbol in protein:\n",
      "            alphabet = Alphabet.HasStopCodon(alphabet, stop_symbol)\n",
      "\n",
      "        return Seq(protein, alphabet)\n",
      "\n",
      "    def ungap(self, gap=None):\n",
      "        \"\"\"Return a copy of the sequence without the gap character(s).\n",
      "\n",
      "        The gap character can be specified in two ways - either as an explicit\n",
      "        argument, or via the sequence's alphabet. For example:\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import generic_dna\n",
      "        >>> my_dna = Seq(\"-ATA--TGAAAT-TTGAAAA\", generic_dna)\n",
      "        >>> my_dna\n",
      "        Seq('-ATA--TGAAAT-TTGAAAA', DNAAlphabet())\n",
      "        >>> my_dna.ungap(\"-\")\n",
      "        Seq('ATATGAAATTTGAAAA', DNAAlphabet())\n",
      "\n",
      "        If the gap character is not given as an argument, it will be taken from\n",
      "        the sequence's alphabet (if defined). Notice that the returned\n",
      "        sequence's alphabet is adjusted since it no longer requires a gapped\n",
      "        alphabet:\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import IUPAC, Gapped, HasStopCodon\n",
      "        >>> my_pro = Seq(\"MVVLE=AD*\", HasStopCodon(Gapped(IUPAC.protein, \"=\")))\n",
      "        >>> my_pro\n",
      "        Seq('MVVLE=AD*', HasStopCodon(Gapped(IUPACProtein(), '='), '*'))\n",
      "        >>> my_pro.ungap()\n",
      "        Seq('MVVLEAD*', HasStopCodon(IUPACProtein(), '*'))\n",
      "\n",
      "        Or, with a simpler gapped DNA example:\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import IUPAC, Gapped\n",
      "        >>> my_seq = Seq(\"CGGGTAG=AAAAAA\", Gapped(IUPAC.unambiguous_dna, \"=\"))\n",
      "        >>> my_seq\n",
      "        Seq('CGGGTAG=AAAAAA', Gapped(IUPACUnambiguousDNA(), '='))\n",
      "        >>> my_seq.ungap()\n",
      "        Seq('CGGGTAGAAAAAA', IUPACUnambiguousDNA())\n",
      "\n",
      "        As long as it is consistent with the alphabet, although it is\n",
      "        redundant, you can still supply the gap character as an argument to\n",
      "        this method:\n",
      "\n",
      "        >>> my_seq\n",
      "        Seq('CGGGTAG=AAAAAA', Gapped(IUPACUnambiguousDNA(), '='))\n",
      "        >>> my_seq.ungap(\"=\")\n",
      "        Seq('CGGGTAGAAAAAA', IUPACUnambiguousDNA())\n",
      "\n",
      "        However, if the gap character given as the argument disagrees with that\n",
      "        declared in the alphabet, an exception is raised:\n",
      "\n",
      "        >>> my_seq\n",
      "        Seq('CGGGTAG=AAAAAA', Gapped(IUPACUnambiguousDNA(), '='))\n",
      "        >>> my_seq.ungap(\"-\")\n",
      "        Traceback (most recent call last):\n",
      "           ...\n",
      "        ValueError: Gap '-' does not match '=' from alphabet\n",
      "\n",
      "        Finally, if a gap character is not supplied, and the alphabet does not\n",
      "        define one, an exception is raised:\n",
      "\n",
      "        >>> from Bio.Seq import Seq\n",
      "        >>> from Bio.Alphabet import generic_dna\n",
      "        >>> my_dna = Seq(\"ATA--TGAAAT-TTGAAAA\", generic_dna)\n",
      "        >>> my_dna\n",
      "        Seq('ATA--TGAAAT-TTGAAAA', DNAAlphabet())\n",
      "        >>> my_dna.ungap()\n",
      "        Traceback (most recent call last):\n",
      "           ...\n",
      "        ValueError: Gap character not given and not defined in alphabet\n",
      "\n",
      "        \"\"\"\n",
      "        if hasattr(self.alphabet, \"gap_char\"):\n",
      "            if not gap:\n",
      "                gap = self.alphabet.gap_char\n",
      "            elif gap != self.alphabet.gap_char:\n",
      "                raise ValueError(\n",
      "                    \"Gap {0!r} does not match {1!r} from alphabet\".format(\n",
      "                        gap, self.alphabet.gap_char\n",
      "                    )\n",
      "                )\n",
      "            alpha = Alphabet._ungap(self.alphabet)\n",
      "        elif not gap:\n",
      "            raise ValueError(\"Gap character not given and not defined in alphabet\")\n",
      "        else:\n",
      "            alpha = self.alphabet  # modify!\n",
      "        if len(gap) != 1 or not isinstance(gap, str):\n",
      "            raise ValueError(\"Unexpected gap character, {0!r}\".format(gap))\n",
      "        return Seq(str(self).replace(gap, \"\"), alpha)\n",
      "\n",
      "    def join(self, other):\n",
      "        \"\"\"Return a merge of the sequences in other, spaced by the sequence from self.\n",
      "\n",
      "        Accepts all Seq objects and Strings as objects to be concatenated with the spacer\n",
      "\n",
      "        >>> concatenated = Seq('NNNNN').join([Seq(\"AAA\"), Seq(\"TTT\"), Seq(\"PPP\")])\n",
      "        >>> concatenated\n",
      "        Seq('AAANNNNNTTTNNNNNPPP')\n",
      "\n",
      "        Throws error if other is not an iterable and if objects inside of the iterable\n",
      "        are not Seq or String objects\n",
      "        \"\"\"\n",
      "        if not isinstance(other, _Iterable):  # doesn't detect single strings\n",
      "            raise ValueError(\"Input must be an iterable\")\n",
      "        if isinstance(other, basestring):\n",
      "            raise ValueError(\"Input must be an iterable\")\n",
      "        from Bio.SeqRecord import SeqRecord  # Lazy to avoid circular imports\n",
      "\n",
      "        a = self.alphabet\n",
      "        for c in other:\n",
      "            if isinstance(c, SeqRecord):\n",
      "                raise TypeError(\"Iterable cannot contain SeqRecords\")\n",
      "            elif hasattr(c, \"alphabet\"):\n",
      "                if a != c.alphabet:\n",
      "                    if not Alphabet._check_type_compatible([a, c.alphabet]):\n",
      "                        raise TypeError(\n",
      "                            \"Incompatible alphabets {0!r} and {1!r}\".format(\n",
      "                                a, c.alphabet\n",
      "                            )\n",
      "                        )\n",
      "                    a = Alphabet._consensus_alphabet([a, c.alphabet])\n",
      "            elif not isinstance(c, basestring):\n",
      "                raise ValueError(\"Input must be an iterable of Seqs or Strings\")\n",
      "        temp_data = str(self).join([str(z) for z in other])\n",
      "        return self.__class__(temp_data, a)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(inspect.getsource(Bio.Seq.Seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TTATACCAT'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reverse_complement(dnaseq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
